ui/src/main/java/edu/wpi/grip/ui/codegeneration/data/TPipeline.java

```diff
@@ -0,0 +1,191 @@
+
+import edu.wpi.grip.core.Connection;
+import edu.wpi.grip.core.Pipeline;
+import edu.wpi.grip.core.Step;
+import edu.wpi.grip.core.sockets.InputSocket;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.ui.codegeneration.TemplateMethods;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * TPipeline(template pipeline) is a data structure that holds the information about a pipeline
+ * needed by the velocity templates to generate code.
+ */
+public class TPipeline {
+
+  protected List<TStep> steps;
+  private int numSources;
+  private int numOutputs;
+  private Map<InputSocket, TOutput> connections;
+  private Map<String, Integer> uniqueSources;
+
+
+  /**
+   * Creates a Tpipeline from a pipeline
+   *
+   * @param pipeline The current grip pipeline.
+   */
+  public TPipeline(Pipeline pipeline) {
+    this.uniqueSources = new HashMap<>();
+    this.steps = new ArrayList<TStep>();
+    this.numSources = 0;
+    this.numOutputs = 0;
+    connections = new HashMap<InputSocket, TOutput>();
+    set(pipeline);
+  }
+
+  /**
+   * sets up the entire Tpipeline by creating the Tsteps
+   *
+   * @param pipeline The grip pipeline used to create the TPipeline.
+   */
+  public void set(Pipeline pipeline) {
+    int count = 0;
+    for (Step step : pipeline.getSteps()) {
+      TStep tStep = new TStep(step.getOperationDescription().name(), count);
+      steps.add(tStep);
+      for (OutputSocket output : step.getOutputSockets()) {
+        TOutput tOutput = new TOutput(TemplateMethods.parseSocketType(output), numOutputs);
+        numOutputs++;
+        tStep.addOutput(tOutput);
+        if (!output.getConnections().isEmpty()) {
+          for (Object con : output.getConnections()) {
+            connections.put(((Connection) con).getInputSocket(), tOutput);
+          }
+        }
+      }
+      count++;
+    }
+    for (int i = 0; i < pipeline.getSteps().size(); i++) {
+      for (InputSocket input : pipeline.getSteps().get(i).getInputSockets()) {
+        TInput tInput;
+        String type = TemplateMethods.parseSocketType(input);
+        if (type.equals("Type")) {
+          type = steps.get(i).name() + "Type";
+        }
+        type = type.replace("Number", "Double");
+        String name = TemplateMethods.parseSocketName(input);
+        if (!input.getConnections().isEmpty()) {
+          if (connections.containsKey(input)) {
+            tInput = new TInput(type, name, connections.get(input));
+          } else {
+            tInput = null;
+            for (Object con : input.getConnections()) {
+              tInput = createInput(type, name, "Connection" + ((Connection) con).getOutputSocket()
+                  .toString());
+            }
+          }
+        } else {
+          tInput = createInput(type, name, TemplateMethods.parseSocketValue(input));
+        }
+        this.steps.get(i).addInput(tInput);
+      }
+    }
+  }
```

Comment:
I feel like the complexity of this step is incredibly high and could be greatly reduced.